opal/util/opal_gethostname.c

```diff
@@ -0,0 +1,173 @@
+ * If there is a preprocessor macro that redefined the call to
+ * gethostname, we undefine that here 
+ */
+#ifdef gethostname
+#undef gethostname
+#endif
+
+/* Build with -DTESTING=1 if we are building small testing standalone (with test_opalgethostname.c) */
+#ifdef TESTING
+#define gethostname(a, b) dummy_gethostname((a), (b))
+#define OPAL_LOCAL_MAXHOSTNAMELEN 4096
+bool first_pass = true;
+
+/* dummy_gethostname function used for testing */
+int dummy_gethostname( char *, size_t );
+
+#else /* this is for a production build, not a standalone test build, so we use the real gethostname() and declare first_pass as static */
+static bool first_pass = true;
+#endif /* ifdef TESTING */
+
+
+#define NUM_TRIES_FOR_NULL_HOSTNAME 8
+
+#define MEMORY_ERROR_STRING "Invalid address (EFAULT) in call to gethostname. Check memory usage."
+#define HOSTNAMELEN_ERROR_STRING "Error in call to gethostname() - check hostname length. Excessively long hostnames are unsupported."
+#define UNKNOWN_ERROR_STRING "Unknown error in opal_gethostname() function. Probable code issue."
+/*
+ * This gethostname wrapper does not return the full-length hostname in
+ * those cases where it is too long for the buffer. It does, however,
+ * guarantee a null-terminated hostname is returned, even if it's
+ * truncated. It also tries again in the case where gethostname returns an 
+ * error because the buffer is too short, but once it has retrieved the full
+ * name without error, it truncates the result and resets errno to 
+ * ENAMETOOLONG. 
+ */
+static char opal_local_hostname[OPAL_LOCAL_MAXHOSTNAMELEN] =
+    HOSTNAMELEN_ERROR_STRING;
+
+
+const char *opal_gethostname( void )
+{
+    size_t count;
+    size_t length = OPAL_LOCAL_MAXHOSTNAMELEN;
+    int ret_val, num_tries = 0;
+    int temp_errno = 0;
+
+    if( first_pass ) {
+        char *buf = malloc( length );
+        first_pass = false;
+        /*
+         * Terminate the buffer first in the 0th position in case we get a 
+         * null result.
+         */
+        buf[0] = '\0';
+
+        while( 1 ) {
+            /*
+             * Offer all but the last byte of the buffer to gethostname. 
+             */
+            ret_val = gethostname( buf, length - 1 );
+            /*
+             * Terminate the buffer in the last position.
+             */
+            buf[length - 1] = '\0';
+            if( 0 == ret_val ) {
+                count = strlen( buf );
+                /*The result was not truncated */
+                if( count > 0 && count < length - 1 ) {
+                    break;
+                }
+                /*
+                 * The buffer is empty. In some gethostname
+                 * implementations this can be because the buffer was too 
+                 * small. But, if it is zero for some other reason, we
+                 * could get caught in the loop. So, break after
+                 * NUM_TRIES_FOR_NULL_HOSTNAME. 
+                 */
+                else if( 0 == count ) {
```

Comment:
What version of `gethostname()` returns 0 (success) yet has an empty buffer?

Reply:
I am trying to account for all possible cases. 
http://perfec.to/gethostname/
While there is not a case of 0 and an empty buffer listed here, I have no evidence that it's not possible. Certainly it does no harm.

Reply:
There's certainly nothing wrong with trying to account for all cases. I asked this question because in your comment you mentioned "in some implementations..." which made me think there was a case or a version or a version of `gethostname()` where this was possible. If there isn't, maybe it can be tightened up a little bit.

Reply:
Well, I was using this writeup as a guide more than the prior one I pasted: http://www.microhowto.info/howto/determine_the_fully_qualified_hostname_of_the_local_machine_in_c.html

It's a more complicated and lengthier discussion, but the way I read it, it seemed to imply that a non-error return code and a null buffer were both possible in some implementations.